Refactor first chapter: first step, the section on basics.

Zimmi48 opened the pull request:
**Kind:** documentation

Only the last commit belongs to this PR. It introduces the first section of the Core language chapter, as initially planned in #12012.

cc @mattam82 and @jfehrle 

After reviewing, it should be possible to use the trick presented by @cpitclaudel to preserve the history of the section on flags, options and tables, since it was almost left unchanged. However, this kind of things must be done last given the work that it represents and how it makes it harder to take comments into account afterward.

TODO
- [x] Adapt `common.edit_mlg` and `orderedGrammar`
- [x] Restore some history for the parts that were moved (lexical conventions, attributes and options)

jfehrle commented:
You want general feedback on the layout or a close reading (partial or full) of the text?  Hard to tell which parts are just cut&pasted.

Zimmi48 commented:
I'd like a rather close reading if possible. To the exception of the section on options and the section on lexical conventions, a lot of text has been heavily edited.

Zimmi48 commented:
But of course, you may start by layout comments if you have some.

Zimmi48 commented:
> Looks like good progress.
> 
> I wonder if it would make sense to do a copy-editing pass on the text near the end of the review process. I could push wording suggestions to GitHub, which you could further edit or comment on (e.g. if you disagree). It seems tedious for both of us to try to do this strictly through comments. Of course, we'd want to get in sync on what's worth rewording versus merely personal preferences. WDYT?

Sounds like a good idea.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+Lexical conventions
+~~~~~~~~~~~~~~~~~~~
+
+Blanks
+  Space, newline and horizontal tab are considered blanks.  Blanks are
+  ignored but they separate tokens.
```
Oh, that's strange! My local build looks like yours.

BTW, I've now got the same build error as you got locally. Cf. #12192. I'll work around them locally using `SKIP_COQTOP` as defined in https://github.com/coq/coq/issues/10901#issuecomment-620508881.

jfehrle replied:
I got the build to work.  Any thoughts on the difference in HTML appearance?  Would be good to get those in sync.

Zimmi48 replied:
Yes, the issue has been fixed now.

Zimmi48 commented on doc/sphinx/language/core/index.rst:
```diff
@@ -6,30 +6,34 @@ Core language
+This separation between the kernel on the one hand and the
+:ref:`elaboration engine <extensions>` and :ref:`tactics
+<writing-proofs>` on the other hand follows what is known as the
+:gdef:`de Bruijn criterion` (keeping a small and well delimited
```
Thanks, I've rebased on the latest version.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+   sentence
+
+     Coq documents are made of a series of sentences, which are
+     :term:`commands <command>` or :term:`tactics <tactic>`, generally
```
OK, I've reintroduced your sentence about simple and compound tactics but below the grammar (contrary to the description of the `sentence` term which is now above).

jfehrle replied:
OK, we can tweak the description for `sentence` later.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+        avoids confusing a space-separated list of terms or
+        identifiers with a :n:`@term_application`.
+
+   type
+
+     Types are a sub-class of the syntactical class of terms:
```
> Think type and term should stay in Gallina spec language chapter.

Remember that the goal is to split this chapter apart and eventually remove it entirely. Furthermore, the new structure will alternate between description of term constructs and of commands. For instance, there is no point in describing the `match` constructs in term before having described the `Variant` command, nor the `fix` or `cofix` constructs before having described the `Inductive` and `CoInductive` commands.

> What does "syntactical class" add here? Is this used to include both well-typed and non-well-typed terms when "term" will generally be used to refer to well-type terms?
> 
> "class" itself may deserve a definition. Perhaps it's used in the literature rather than the common term "set"?

You are right that this is not very clear. Well-typed terms are also a subset of syntactically valid terms. The goal here was just to say that we cannot distinguish types purely on their syntax. I'll reformulate.

jfehrle replied:
Let me try again:
- the entries for "sentence" .. "attribute" are closely related and are presented in a logical reading order
- the entries for "term" and "type" are closely related and should appear together, but have little relationship to "sentence"...  It introduces many nonterminals that are not described at all here. 
 So it's a disconnect going from term/type to sentence...  I don't care where they end up, but _they don't belong here_.
- sentence... begins the top-down description of the grammar--an obvious entry point for the reader

Are you expecting to add a lot more glossary terms here?  I think that will not be very readable.  I expect you can have multiple `.. glossary::` directives in the document--no need to jam things together here.

Zimmi48 replied:
Terms, types, tactics and commands are the most essential Coq notions. Never mind that they are independent of one another, what is the problem of packing them in the same initial file? It has a logical meaning, which is that this is the base vocabulary you need to understand to go further.
Creating an arbitrary split between the notions of terms on the one hand, and commands on the other, was exactly what the Gallina specification language chapter did, and what I want to move away from because of what I said on `match` making no sense if you didn't describe `Variant` first, etc.

Zimmi48 replied:
If you think this is a bad idea, I'd like to know ASAP. In fact, I would have hoped to know from #12012 only, but I understand that this was a bit too abstract to comment on.

jfehrle replied:
> Terms, types, tactics and commands are the most essential Coq notions. Never mind that they are independent of one another, what is the problem of packing them in the same initial file?

Why not add something like the first sentence before the glossary?  As it is, the glossary seems a bit like a laundry list because the relationship between sentence and type is unclear/unstated.  We should be extra clear in this introductory material.

Also, there's no explanation of what terms and types are used for (unlike the others).

The long grammar for term is out of place; _none_ of it is explained here.  Command is fairly well described though it could be improved, e.g. doesn't cover selectors.

I think sentence... followed most logically after explaining lexical conventions.  At the least I'd put them at the top.

Zimmi48 replied:
I'll try to add much more explanations. I have to agree that the current state is not really readable.

Zimmi48 replied:
Check what I wrote.

jfehrle replied:
It's a little better.  I still think there's too much syntax for term here.  But this is not where and how we should introduce _term_ and especially _type_.  We don't yet have a proper high-level introduction to term and type and the relevant points from theory.  For example, that type checking is how you verify a proof and that type checking is just 2nd order unification.  The Curry-Howard correspondence. Also how CoC compares to other logical foundations.  Then this section can just focus on syntax, which seems more consistent with sentence...

This should go in the introduction chapter.  We should add this fairly soon.

We can organize it as it is for the time being and revisit this point when we have that introduction.

JasonGross commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+Syntax conventions
+~~~~~~~~~~~~~~~~~~
+
+The grammar of Coq is an extensible, non-context-free grammar.  The
```
I think it's a stretch to call it context-free at any given point.  A context-free presentation would have the grammar be modified at every single `Definition`, because you can bind a scope to a type so that scope is automatically opened for arguments of that to for any definition, and you can define a scope to automatically open a new custom entry with completely separate grammar rules.  So your CFG would need to have a separate rule for every definition in scope.

jfehrle replied:
Simple lookahead doesn't make the grammar non context free.

Zimmi48 replied:
No more reference to context-free grammars.

JasonGross commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+  Hangul, Hiragana and Katakana characters, CJK ideographs,
+  mathematical letter-like symbols and non-breaking
+  space. :production:`unicode_id_part` non-exhaustively includes
+  symbols for prime letters and subscripts.
```
Seems strange to me to include subscripts but not superscripts, even in a non-exhaustive list

Zimmi48 replied:
Do you know in which list they should appear? I don't know much about the Unicode support.

JasonGross replied:
Pretty sure subscripts and superscripts should appear in the same list, whichever the list that is.  I'd believe Unicode id part, but I'm not sure.

Zimmi48 replied:
I wonder what "prime letters" mean.

jfehrle replied:
Perhaps they are referring to prime characters in Unicode, which has characters for single, double triple prime, both forward and backward.

Subscripts work, but superscripts are not accepted:

![image](https://user-images.githubusercontent.com/1253341/80636322-21d6ef80-8a12-11ea-92f9-99e248a6f1a2.png)

Gives `Syntax Error: Lexer: Undefined token`.  And notice the subscripts are not all aligned the same way.  (This is in CoqIde.)

Look at unicode.ml and the very dull unicodetable.ml.

JasonGross replied:
Oooh, is this because superscripts are considered to be exponent-like things, where as subscripts are part of identifiers?

Zimmi48 replied:
OK, I can revert this change.

JasonGross commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+  The set of tokens defined at any given time can vary because the
+  :cmd:`Notation` command can define new tokens.  A :cmd:`Require`
+  command may load more notation definitions, while the end of a
+  :cmd:`Section` may remove notations.  Some notations are defined in
```
Do we have a `Remove Notation` command yet?  If so, it should be mentioned here.

Zimmi48 replied:
I don't think any such command exists.

JasonGross commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+  loaded automatically at startup time.
+
+  Here are the character sequences that Coq directly defines as tokens
+  without using :cmd:`Notation`::
```
Aren't some of these defined in the prelude using `Notation`?  For example, <= = ->

Zimmi48 replied:
I thought @jfehrle has done a pass on these symbols to provide an accurate list. To confirm with him.

Zimmi48 replied:
Thanks for your answer! So AFAICT there is nothing to change here. (Although of course, this list will not be kept automatically in sync, contrary to the grammar snippets.)

jfehrle replied:
FWIW: unlike the grammar snippets

JasonGross commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+
+     .. note::
+
+        Many commands and tactics use :n:`@one_term` rather than
```
Why not `@term1`?

Zimmi48 replied:
Not sure I understand your question, but `one_term` additionally supports `@foo`. Note that `one_term` corresponds (IIRC) to `constr` in the implementation, while `term` corresponds to `lconstr`.

JasonGross replied:
Sorry, I guess my question is actually "why are one_term and term1 different nonterminals? What would go wrong of we inlined term1 into one_term and then did s/one_term/term1/ everywhere?"

Zimmi48 replied:
Oh! That's an interesting consideration. The syntax that is shown here is simplified (in truth `term10` is not defined in terms of `one_term`) but maybe that could be done.

Zimmi48 replied:
The problem is that there is an intermediate precedence level (that is not shown in the doc) that actually distinguishes further the two.

jfehrle replied:
We have:

```
term10: [
| term1 LIST1 arg
| "@" qualid OPT univ_annot LIST0 term1
| term1
]

one_term: [
| term1
| "@" qualid OPT univ_annot
]

term1: [
| term_projection
| term0 "%" scope_key
| term0
]
```

Combining one_term with one or the other would change the accepted language.  Perhaps the mlg grammar could be simplified.

Zimmi48 replied:
The change I did of inserting `one_term` into the `term` grammar is equivalent from the perspective of what we display in the documentation, but not in terms of the parser because there are intermediate levels that we don't cover. It will be easier to check when I push the changes to `common.edit_mlg`.

Zimmi48 replied:
Done.

JasonGross commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+     Formally, this gives rise to logical foundations which are not
+     the same as the well-known `"set theory"
+     <https://en.wikipedia.org/wiki/Set_theory>`_.  Coq is based on
+     the Calculus of Inductive Construction which is a `"type theory"
```
I thought it was "Constructions" not "Construction"?

Zimmi48 replied:
That's a typo, thanks.

JasonGross commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+
+     Terms are syntactically valid objects constructed from the
+     following grammar.  When they are :term:`well-typed`, they
+     represent mathematical objects.
```
You can work with untyped terms in Ltac using `uconstr`.  Or you can fake well-typedness by cheating with `exact_no_check`.  But insofar as there's a map from syntax to semantics (formally, from Coq ASTs to sets or types or classes or points in a type), only the well-typed terms get mapped to something.

Zimmi48 replied:
Right. I didn't want to write just "they represent mathematical objects" because `0 1` is syntactically valid and not well-typed and does not carry any meaning.

jfehrle replied:
How about something more like this:

```
Definitions, theorems, proofs and other constructs are represented in |Coq| as *terms*.
Terms are constructed from the following grammar.  Generally, |Coq| works with
terms that are well-typed.
```

We should define open_term (aka uconstr) somewhere.  Could be here or elsewhere.

jfehrle replied:
"mathematical objects" would need to be defined.  It's better to avoid introducing this term.

Zimmi48 replied:
OK, I agree that my text was not substantial enough. Will change to look more like what you suggest. Regarding `open_term`, this is really only useful to explain for people who write tactics, so to be left for the Ltac chapter (especially as we will move the Tactic Notation stuff over there).

Zimmi48 replied:
Check what I wrote.

JasonGross commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+   tactic
+
+     Tactics specify how to transform the current proof state as a
+     step in creating a proof.  They are syntactically valid only when
```
Is it actually a syntax error to use these when not in proof mode?  If so, this gives another example of non-context-freeness in conjunction with the deprecated flag Refine Instance Mode, IIRC

Zimmi48 replied:
Yes. But to me this is an example of non-context-freeness even without things like `Refine Instance Mode` (which has been removed BTW). I mention it already in the first paragraph on syntax:

>Furthermore, some parsing rules are only activated in certain contexts (interactive proof mode, custom entries...).

JasonGross commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+
+.. cmd:: Add @setting_name {+ {| @qualid | @string } }
+
+   Adds the specified values to the table :n:`@setting_name`.
```
How does `Add Warnings Append` fit into this?  Is "Append" part of the setting name?

Zimmi48 replied:
I think you mean `Set Warnings Append` (which is a strict equivalent of `Set Warnings` BTW). The answer is that this is the black sheep. Since a recent PR of mine (#11997), it is the last remaining user of `declare_string_option` and the only option that does not support the `Unset` command.

JasonGross commented on doc/sphinx/language/extensions/implicit-arguments.rst:
```diff
@@ -3,6 +3,23 @@
+
+This is a :term:`bar name` reference.  And another: :term:`bar name`
+
+This is a :term:`fizname` reference.
```
I'm confused by this addition.  Why is it in the implicit arguments section?

Zimmi48 replied:
That's a test commit coming from the glossary PR. It won't be there anymore in my next push.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+
+The grammar of Coq is an extensible, non-context-free grammar.  The
+extensions to the Coq grammar can come from either plugins, or the
+:ref:`notation system <syntaxextensionsandinterpretationscopes>`.
```
Yes, both dashes and underscores work.

JasonGross replied:
In that case, why not break up such long names with dashes or underscores?

Zimmi48 replied:
Yeah, I've done it for this example. I'll push in a moment.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+Basic notions
+-------------
+
+Glossary
```
It is there because there is another sub-section at this level (Flags, Options and Tables) and I want the TOC that appears in the left-bar to be balanced.

jfehrle replied:
Why does it have to be balanced?  What difference does balancing make?  The title "Glossary" suggests like it's the one-and-only glossary in the document.

Zimmi48 replied:
Let's change the title then. But having only one sub-section in a section does not provide an accurate idea of the content of the section when looking at the TOC.

jfehrle replied:
You could make Flags, Options and Table a section.

I still think terms and types doesn't belong here--they're the essence of the unique functionality of Coq.  The other things in this chapter are the sort of very routine things any language manual has to cover.  Of course if you've used several proof checkers, then types and terms are routine, too, but I don't think that applies to most of our users.

Zimmi48 replied:
Indeed, there are good chances that a majority of our users have not used other proof assistants, however, all programming languages have "expressions" which are basically the same things as "terms" and a lot of them have "types" as well (except that they usually form a distinct class of objects.

JasonGross commented on doc/sphinx/language/gallina-specification-language.rst:
```diff
@@ -15,188 +15,17 @@ language of commands, called *The Vernacular* is described in Section
-Syntax of terms
----------------
-
 The following grammars describe the basic syntax of the terms of the
```
I think "grammars" is a valid plural in technical English, at least.  Unless you're saying that it doesn't match the section?

Zimmi48 replied:
Let me remove this outdated paragraph entirely.

jfehrle replied:
> I think "grammars" is a valid plural in technical English, at least.

Sure, the word can be plural (the C and Java grammars) but subsets of the productions in a single grammar, e.g. Coq, are usually not referred to as grammars.

Zimmi48 replied:
FTR, I've removed this paragraph entirely. But to add fuel to the discussion, I guess the right term here should be "the grammar snippets".

jfehrle replied:
FWIW "snippets" sounds a bit casual and imprecise.  Maybe just "productions"--if that conveys the right idea.  I don't recall the exact wording that you had before.

Zimmi48 replied:
That doesn't really matter though, we are discussing a paragraph that is not there anymore.

JasonGross commented:
Question: should we take the opportunity here to be consistent about using `|Coq|` rather than `Coq`?  (Do they even display differently?)

Zimmi48 commented:
> Question: should we take the opportunity here to be consistent about using `|Coq|` rather than `Coq`? (Do they even display differently?)

Absolutely, thanks for reminding me (cf. #11924).

JasonGross commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+
+     .. warn:: Unsupported attribute
+
+        This warning is fatal by default.  It is caused by using a
```
There's a difference between "makes coqc exit" and "makes coqtop/CoqIDE exit".  I suspect @jfehrle was talking about the latter?  Where in the manual do we talk about the ways of interacting with Coq?  It does seem plausible that we should talk about ways of interacting with Coq and warnings and errors pretty early on

Zimmi48 replied:
It's in the "Using" part, more precisely the "Command-line and graphical tools" chapter, so the last part in fact.

jfehrle replied:
You could just say "this warning is configured to behave as an error by default.  You can change that by modifying :opt:`Warnings`."

OTOH, Warning is unusable:
- the warning names/categories are undocumented
- no error checking of the option value `Warnings "garbage"` is accepted without complaint

Zimmi48 replied:
Yes, I know the warning configuration mechanism is a mess. See also my answer to Jason [here](https://github.com/coq/coq/pull/12172#discussion_r416605833).

jfehrle commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+Flags, Options and Tables
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Besides the use of :term:`attributes <attribute>` to control the
+behavior of a single command, it is possible to control the behavior
+of all the following commands using flags, options and tables:
```
I modified the suggestion, take another look.

"it is possible to control the behavior of all the following commands [Set, Unset, Add, Remove, Test, ...] using flags, options and tables" is backwards; those commands are the means for controlling the flags, options and tables.

Zimmi48 replied:
Thanks, I hadn't realized that "following commands" could be confusing. It was meant to designate the commands that follow a `Set` within a Coq document. I still find your suggestion too imprecise but I will improve the wording thanks to your feedback.

jfehrle commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+
+* no attribute: the original setting is *not* restored at the end of
+  the current module or section.
+* :attr:`local` (an alternative syntax is to use the ``Local``
```
Fair enough, I updated the suggestion (wordsmithing only).

jfehrle commented on doc/sphinx/using/libraries/writing.rst:
```diff
@@ -0,0 +1,47 @@
+Deprecating library objects or tactics
+--------------------------------------
+
+You may use the following :term:`attribute` to deprecate a notation or
+tactic.  When renaming a definition or theorem, you can introduce a
```
Perfect!

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,462 @@
+     “:math:`x:T`”.  Informally, “:math:`x:T`” can be thought as
+     “:math:`x` *belongs to* :math:`T`”.
+
+     Types themselves are a subset of terms, but they cannot be
+     distinguished purely on their syntax:[#types]_
```
Regarding the footnote, it is also possible to have them in the middle of the page. Do you think that would be worth? [Usually footnotes are expected to be at the bottom of the page though.]

jfehrle replied:
The footnote option is one setting that applies to all footnotes throughout the document?  Or can you choose?  So many of our pages are so very long.  I think it's probably good if the footnotes are closer--people are more likely to stumble across them and learn something useful they might not have planned on learning.

Zimmi48 replied:
You can choose where you put each footnote. It's just about where you put them in the file. I suppose we can consistently decide to always put them at the end of the current section or sub-section.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,462 @@
+     .. prodn::
+        type ::= @term
+
+     The kernel of Coq is a type checker that verifies that a term has
```
"is". Note that this is something that I repeat from the Core language introduction:
> The kernel is a type checker which verifies that terms have their expected type.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,462 @@
+Syntax conventions
+~~~~~~~~~~~~~~~~~~
+
+Coq uses an extensible, `context-sensitive grammar
+<https://en.wikipedia.org/wiki/Context-sensitive_grammar>`_.  The
```
I've removed all mentions of context-free / context-sensitive and moved part of your note on `Print Grammar` here.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,462 @@
+extensions to the Coq grammar can come from either plugins, or the
+:ref:`notation system <syntax-extensions-and-notation-scopes>`.
+Furthermore, some parsing rules are only activated in certain contexts
+(interactive proof mode, custom entries...).  Nonetheless, the
+grammars are presented in this manual as if they were `context-free
+<https://en.wikipedia.org/wiki/Context-free_grammar>`_.
```
Check what I wrote.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+================================
+Basic vocabulary and conventions
```
FWIW I've slightly changed the title.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,428 @@
+appending a number to the name of the nonterminal (as in :n:`@term100`
+or :n:`@ltac_expr3`).  Precedence levels which are unused by the base
+grammar are not represented, even though they could still be populated
+by notations or plugins.
```
Done.

Zimmi48 commented on doc/sphinx/language/gallina-specification-language.rst:
```diff
@@ -15,188 +15,17 @@ language of commands, called *The Vernacular* is described in Section
 In Coq, logical objects are typed to ensure their logical correctness.  The
```
No more mathematical object in my own text.

Zimmi48 commented:
I feel like this section has much improved thanks to your comments. To me, this looks ready, although we might want to move the explanations on managing warnings here as well (but this could be done as a later step). I'll wait for your final approval before squashing and trying to restore some history regarding the parts that were moved (lexical conventions, attributes and options).

jfehrle commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,509 @@
+
+   type
+
+     To be valid and accepted by the |Coq| kernel, a term needs an
+     associated type.  We express this relationship by “:math:`x` *of
+     type* :math:`T`”, which we write as “:math:`x:T`”.  Informally,
+     “:math:`x:T`” can be thought as “:math:`x` *belongs to*
+     :math:`T`”.
+
+     A specificity of the |Coq| language is that types can depend on
+     terms (we say that the language is `dependently-typed
+     <https://en.wikipedia.org/wiki/Dependent_type>`_).  Because of
+     this, types and terms share a common syntax, all types are terms,
+     but not all terms are types:[#types]_
+
+     .. insertprodn type type
+
+     .. prodn::
+        type ::= @term
+
+     The kernel of |Coq| is a type checker that verifies that a term has
+     the expected type by applying a set of typing rules (see
+     :ref:`Typing-rules`).  If that's indeed the case, we say that the
+     term is :gdef:`well-typed`.
```
How about presenting the essential info in the following order (if correct):
- every type is a term but not every term is a type
- there is a set of typing rules that can determine the type [a type?] associated with most but not all terms
- terms for which a type is determined are considered well-typed
- Coq generally works with well-typed terms

Mentioning the notation could go in the introduction

Dependent types deserve more explanation (elsewhere) and worth our own definition, not just a Wikipedia link.  Not sure they need to be mentioned here.  "A specificity of" misuses the word.  One could say "In the |Coq| language, types can depend ..."

The footnote didn't seem helpful for the beginner--typing rules and sorts have not been explained yet--and maybe too mundane/out of place for non-beginners.

> The kernel of |Coq| is a type checker that verifies that a term has
     the expected type by applying a set of typing rules (see
     :ref:`Typing-rules`).  If that's indeed the case, we say that the
     term is :gdef:`well-typed`.

I thought the definition is that a term is well-typed if the type can be computed--not that a pair (term, type) is consistent.  It may be true and useful that the Coq kernel checks these pairs but that's not the definition or very relevant right here, right?

Some of this could eventually go into the introduction and with more elaboration.

Zimmi48 replied:
>How about presenting the essential info in the following order (if correct):

I am not convinced by your proposed order. However, I've still moved up the paragraph defining well-typed. Have a look and let me know if there is something specific that you don't like in the new order.

>Dependent types deserve more explanation (elsewhere) and worth our own definition, not just a Wikipedia link. Not sure they need to be mentioned here. "A specificity of" misuses the word. One could say "In the |Coq| language, types can depend ..."

Sure, they can be explained further when we reach the notion of product types. At that point, we can introduce a glossary definition, and then we can link to it from there instead of linking to Wikipedia. But my reason for mentioning dependent types at that point was to justify that types and terms share a common language, whereas this is not usually the case in standard typed programming languages.

>"A specificity of" misuses the word. One could say "In the |Coq| language, types can depend ..."

I'm replacing "A specificity of" by "A special feature of". I didn't like your suggestion because it removed the aspect that this is something that differs from the other languages, which is precisely what I wanted to emphasize.

>The footnote didn't seem helpful for the beginner--typing rules and sorts have not been explained yet--and maybe too mundane/out of place for non-beginners.

I introduced the footnote because I thought you said that the statement "all types are terms, but not all terms are types" is too imprecise. I'm removing it.

---

To answer the rest of your comments:

>most but not all terms

Saying that we can find types for "most" terms does not make any sense. There is a countable number of terms that is divided into a countable number of well-typed terms and a countable number of non-well-typed terms. If we were to use another metric than cardinality (like probability to build a well-typed term), I would intuitively say that well-typed terms are completely outnumbered by non-well-typed-terms.

>Coq generally works with well-typed terms

Generally would be too vague. It should be "virtually always" but then people would wonder what are the exceptions. This is not useful info at this point. That's why I just wrote "To be valid". The only thing that I wanted to express is that there are things that you can construct with the term grammar presented above which won't be valid terms because they won't be well-typed.

>I thought the definition is that a term is well-typed if the type can be computed--not that a pair (term, type) is consistent.

You can have well-typed terms for which the type cannot be computed but if you do provide the type, it can be checked. Type inference is not complete in Coq contrary to a simpler type system like OCaml's. Probably something else to mention somewhere, for example whenever we do get to the point of defining the terms "type inference" and "type checking".

jfehrle replied:
Oh, checking is decidable?  Good to know.  Eager to see this info in the doc.

jfehrle replied:
> I am not convinced by your proposed order.

It's best to begin with a positive statement about types and then to qualify it.  That's why I'd put "[syntactically] all types are terms" first, after which "but not all terms are types" follows naturally.  Then explain why not all terms are types.

> Saying that we can find types for "most" terms does not make any sense.

Poor wording by me.  Should have been for "some" types.

> Generally would be too vague.

Not at all.  "Generally" means "in general terms; without regard to particulars or exceptions". It could mean 51% of the time or 99.99% of the time.  It implies that there are exceptions but that they're not important enough to mention here.

Zimmi48 replied:
>Oh, checking is decidable? Good to know. Eager to see this info in the doc.

It is, unless you activate some unsafe flags like [`Cumulative StrictProp`](https://coq.github.io/doc/master/refman/addendum/sprop.html#coq:flag.cumulative-strictprop).

jfehrle commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,509 @@
+=============================
+Basic notions and conventions
```
Good enough for now.  Maybe later on, as the document evolves, a better title will be obvious.

jfehrle commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,509 @@
+Syntax conventions
+~~~~~~~~~~~~~~~~~~
+
+|Coq| contains an extensible parsing engine.  The extensions to the
+|Coq| syntax can come from either plugins, or the :ref:`notation
+system <syntax-extensions-and-notation-scopes>`.  Furthermore, some
+parsing rules are only activated in certain contexts
+(:ref:`interactive proof mode <proofhandling>`, :ref:`custom entries
+<custom-entries>`...).  The syntax described in this documentation is
+equivalent to the rules implemented in the |Coq| parser, but has been
+heavily edited to improve readability and presentation.
```
Think I'd reorder the sentences something like this (with a little rewording):

The syntax described in this documentation is
equivalent that accepted by the |Coq| parser, but the grammar has been
heavily edited to improve readability and presentation.

Coq uses an extensible parser. Plugins and the ref:`notation
system <syntax-extensions-and-notation-scopes>` can extend the syntax at run time.
Some notation are predefined by default.  The documented grammar doesn't
include such syntax extensions.

> Furthermore, some
parsing rules are only activated in certain contexts
(:ref:`interactive proof mode <proofhandling>`, :ref:`custom entries
<custom-entries>`...).

Yeah, some things are semantically restricted.  Do you think we need say that
explicitly here?  We've not introduced interactive proof mode or custom entries
and they seem too detailed for here.

WDYT?

jfehrle replied:
Looks good.

jfehrle commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,509 @@
+boldface label "Attribute:".  Attributes are listed in the
+:ref:`attribute_index`.
+
+.. warn:: This command does not support this attribute: @ident.
+   :name: This command does not support this attribute
+
+   This warning is configured to behave as an error by default.  You
+   may change this using the :opt:`Warnings` option:
+
+   .. coqtop:: none
+
+      Set Silent.
+
+   .. coqtop:: all warn
+
+      Set Warnings "unsupported-attributes".
+      #[ foo ] Comments.
```
Great example, very clear!

jfehrle commented:
Hmm, when I clicked "submit review" it said I had 11 comments, but I only see 3 above.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,509 @@
+     the same as the well-known `"set theory"
+     <https://en.wikipedia.org/wiki/Set_theory>`_.  |Coq| is based on
+     the Calculus of Inductive Constructions which is a `"type theory"
+     <https://en.wikipedia.org/wiki/Type_theory>`_.
```
I've removed the phrasing "this gives rise" and mixed our two wordings into something new. I've kept "Formally," because it answers to "Intuitively," in the paragraph above.

Zimmi48 commented:
> Feel free to squash, then I'll look at posting some wordsmithing suggestions as an additional commit in the PR.

I'll squash after you push your own commit since I need to rewrite the history afterward to try to preserve as much as possible.

Do not forget to approve the PR if it now looks in a mergeable state (besides the history rewriting).

jfehrle commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,509 @@
+        | @term10
+        term10 ::= @term_application
+        | @one_term
+        one_term ::= @ @qualid_annotated
```
An index of nonterminals is easy to produce.  doc_grammar could also generate a list of symbols to go in an index (since many symbols are used in multiple productions, they'd have to be qualified by the name of the nonterminal using the symbol).  I could do that after we're done updating syntax in the doc.

OTOH, repeating the grammar in the appendix is almost as good, less work and helpful in some other ways.  Users can use the browser search to find things.

Zimmi48 replied:
You might be right (but then the repeated productions would have to link back to the place where they are documented). OTOH having an index (with the qualified disambiguation like you said) would show instantly to the user that there are several rules corresponding to same terminal (typical example: `match`). Whereas some users might search for `match` in the full grammar and not realize that the first instance is not the only occurrence.

jfehrle replied:
My browser, Chrome, shows the number of matches whenever you do a text search, so it's easy to see that there are multiple occurrences.  I don't know if most browsers do that.

Zimmi48 replied:
I think so. But it is easy to miss if you don't realize that there might be several matches.

jfehrle commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,509 @@
+
+     The base expressions of the |Coq| language are terms.  Terms can
+     represent computable programs, program types, mathematical
+     propositions and mathematical proofs.
```
> OK for the reordering, but Coq is not just for mathematicians, it is also for programmers / engineers who want to verify software that run. So I'm keeping the mention of programs.

I wasn't suggesting to remove program but just to move them in the list.  Indeed, my interest i in working with programs.  I wanted to make the math folks feel comfortable.

jfehrle replied:
Looks good, BTW.

jfehrle commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,526 @@
+     constructs is presented in a dedicated section.  Some of these
+     constructs (like :n:`@term_forall_or_fun`) are part of the core
+     language that the kernel of |Coq| understands and are therefore
+     described in :ref:`the present chapter <core-language>`, while
+     some others (like :n:`@term_if`) are language extensions that are
+     presented in :ref:`the next chapter <extensions>`.
```
"the present chapter" and "the next chapter are placeholders?  Eventually the first should directly link to the types section, the second should have the title.  You think?

Zimmi48 replied:
No, they're not, it's to remind the distinction between the Core language and the Language extensions chapters that we have made in the general introduction and the introductions of these chapters.

jfehrle commented on doc/sphinx/language/extensions/implicit-arguments.rst:
```diff
@@ -420,6 +420,14 @@ but succeeds in
+.. prodn::
+   term_explicit ::= @ @qualid_annotated
+
+This syntax can be used to disable implicit arguments locally.  To
```
Can we be more specific on the scope than "locally"?  I was thinking something like "within the term" or "with the qualid_annotated", but qualid_annotated can't contain any function applications, so not sure what I'd say.

Zimmi48 replied:
I'll explain better and put an example.

jfehrle commented on doc/sphinx/language/cic.rst:
```diff
@@ -24,7 +24,7 @@ to a type and takes the form “*for all x of type* :math:`T`, :math:`P`”. The
 “:math:`x` *belongs to* :math:`T`”.
 
-The types of types are *sorts*. Types and sorts are themselves terms
+The types of types are :gdef:`sort`\s. Types and sorts are themselves terms
```
BTW, I think I'll add support for <> inside gdef as its' supported in :term: -- to allow for plurals, capitalization, etc.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -55,11 +46,34 @@ graphically using the following kind of blocks:
+
+.. warning::
+
+   Given the complexity of these parsing rules, it would be extremely
+   difficult to create an external program that can properly parse a
+   |Coq| document.  Therefore, tool writers are advised to delegate
+   parsing to |Coq|, by communicating with it, for instance through
+   `SerAPI <https://github.com/ejgallego/coq-serapi>`_.
```
The goal is precisely to be discouraging. We have well-supported solutions, and we want to have people stop attempting creating parsers on their own unless they know exactly what they are doing, or unless they don't care about robustness.

jfehrle replied:
OK.  There's a history of people trying to write their own parsers and a) wasting a bunch of their own time because they don't understand the subtleties and/or b) and this wastes the time of other developers?

Zimmi48 replied:
Yes

Zimmi48 commented:
Thanks for the wording adjustments.

herbelin commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,524 @@
+        type ::= @term
+
+     Intuitively, types may be viewed as sets containing terms.  We
+     say that a type is :gdef:`inhabited` if it contains at least one
+     term (i.e. if we can find a term which is associated with this
+     type).  We call such terms :gdef:`witness`\es.  Note that deciding
+     whether a type is inhabited is `undecidable
+     <https://en.wikipedia.org/wiki/Undecidable_problem>`_.
+
+     Formally, types can be used to construct logical foundations for
+     mathematics alternative to the standard `"set theory"
+     <https://en.wikipedia.org/wiki/Set_theory>`_: we call such
+     logical foundations `"type theories"
+     <https://en.wikipedia.org/wiki/Type_theory>`_.  |Coq| is based on
+     the Calculus of Inductive Constructions, which is a particular
+     instance of type theory.
```
These paragraphs look a bit odd compare to the factual tone of the rest of the chapter. Shouldn't they go to an introductory chapter?

jfehrle replied:
I think so, but that section hasn't been written yet.  I think what is written here is a compromise for the time being.

herbelin commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,524 @@
+
+   type
+
+     To be valid and accepted by the |Coq| kernel, a term needs an
+     associated type.  We express this relationship by “:math:`x` *of
+     type* :math:`T`”, which we write as “:math:`x:T`”.  Informally,
+     “:math:`x:T`” can be thought as “:math:`x` *belongs to*
+     :math:`T`”.
+
+     The |Coq| kernel is a type checker: it verifies that a term has
+     the expected type by applying a set of typing rules (see
+     :ref:`Typing-rules`).  If that's indeed the case, we say that the
+     term is :gdef:`well-typed`.
+
+     A special feature of the |Coq| language is that types can depend
+     on terms (we say that the language is `dependently-typed
+     <https://en.wikipedia.org/wiki/Dependent_type>`_).  Because of
+     this, types and terms share a common syntax.  All types are terms,
+     but not all terms are types:
+
+     .. insertprodn type type
+
+     .. prodn::
+        type ::= @term
```
Is the objective here mainly to justify the `type` terminology. If yes, this can certainly be shortened.

jfehrle replied:
It is in lieu of a proper introduction for types.

herbelin commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,524 @@
+   | {| Polymorphic | Monomorphic }
+   | {| Cumulative | NonCumulative }
+   | Private
+   | Program
```
Doesn't this give a too strong importance to legacy attributes? Wouldn't it be clearer to move them to their own section (doomed to be removed at some time).

Maybe the time is ripe also to apply a script which consistently turns all legacy-syntax attributes into `#[ ]` syntax in existing files??

Zimmi48 replied:
To be honest, they still have quite a strong importance. Given that 8.12 will be the first version of Coq where attributes are really put forward in the documentation, and given how frequently legacy attributes are used in practice (including in the documentation), it seems important to me not to hide them under the carpet.

Legacy attributes may be doomed to eventually disappear, but they do not currently create any particular trouble (except `Export` which couldn't be turned into a proper legacy attribute).

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,524 @@
+:n:`{* item }`
+   An optional list of items.
+
+:n:`{+s item}`
+   A list of one or more items separated by "s" (e.g. :n:`item__1 s item__2 s item__3`).
```
And Jim was suggesting "sep". My issue with this is that separators that are more than two letters long do not fit the tiny circle in which this is displayed, resulting in a less than optimal rendering. What about going back to using an example symbol instead (such as ",")?

jfehrle replied:
I think `s` is better than `,`.  `s` suggests that it is a placeholder for which you can substitute other strings.  `,` does not.  `s` is a bit terse but I think OK here.

Zimmi48 replied:
OK

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,524 @@
+  includes Latin, Greek, Gothic, Cyrillic, Arabic, Hebrew, Georgian,
+  Hangul, Hiragana and Katakana characters, CJK ideographs,
+  mathematical letter-like symbols and non-breaking
+  space. :production:`unicode_id_part` non-exhaustively includes
```
Indeed, and all the more because it is allowed at the *beginning* of an indentifier.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,524 @@
+     int ::= {? - } {+ @digit }
+     num ::= {+ @digit }
+     digit ::= 0 .. 9
+
```
You mean that it's already the case in master? Then, I should fix this part.

Zimmi48 replied:
As far as I can tell, `+1` is still rejected in places which expect a `num`.

proux01 replied:
Indeed, it's not yet in the parser: https://github.com/coq/coq/pull/11703/commits/f7687371270e0b3fd3b61db3f25128c567aa7033

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,524 @@
+    _ Axiom CoFixpoint Definition Fixpoint Hypothesis IF Parameter Prop
+    SProp Set Theorem Type Variable as at by cofix discriminated else
+    end exists exists2 fix for forall fun if in lazymatch let match
+    multimatch return then using where with
```
>discriminated: does not have to be a keyword, it is a mistake.

Yes, but the goal is to provide documentation that is accurate. We can adapt the doc whenever this mistake is fixed. In the meantime, it is important to document the reality since this is what prevents users from writing `Definition discriminated := 0.`.

Zimmi48 replied:
>if then else exists: only in Prelude

`if`, `then` and `else` are keywords even with `-noinit`. `as` as well.

herbelin replied:
About `discriminated`: see #12229 which fixes it.

About `using`, `lazymatch`, `multimatch`, `by`: see #12229 which moves them in the tactic chapters.

About `if`, `then` and `else`: my mistake.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,524 @@
+        | @term_generalizing
+        | @term_ltac
+        | ( @term )
+        qualid_annotated ::= @qualid {? @univ_annot }
```
The goal is to put the syntax of a specific construct close to where this construct is documented. @jfehrle has been advocating for having a summary of the full grammar on some dedicated page. If we did that, we could try putting fewer indirections in this reference grammar.

herbelin replied:
And what is the reason for `forall` and `fun` being together?

Zimmi48 replied:
That's because I intend to document them together. Cf. next (upcoming) PR.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,524 @@
+     boldface label "Command:".  Commands are listed in the
+     :ref:`command_index`.  Example:
+
+     .. cmd:: Comments {* @string }
```
Commands can only be defined once in the manual. I chose `Comments` because it was undocumented so far. Moving the description of `Check` here would cancel our effort to document query commands (supporting `@num:` selectors) together.

Zimmi48 commented on doc/sphinx/language/core/basic.rst:
```diff
@@ -0,0 +1,522 @@
+  The following character sequences are reserved keywords that cannot be
+  used as identifiers::
+
+    _ Axiom CoFixpoint Definition Fixpoint Hypothesis IF Parameter Prop
+    SProp Set Theorem Type Variable as at by cofix discriminated else
+    end exists exists2 fix for forall fun if in lazymatch let match
+    multimatch return then using where with
```
```suggestion
    _ Axiom CoFixpoint Definition Fixpoint Hypothesis Parameter Prop
    SProp Set Theorem Type Variable as at cofix discriminated else
    end fix for forall fun if in let match
    return then where with
```

Zimmi48 commented:
I've addressed all the comments I could in this PR. Regarding the remarks of Hugo and Jim on the style of the description of `type`, maybe some of it can indeed be moved elsewhere, but it will be easier to discuss that when the main bulk of the Core language refactoring is over.

I've also used the git octopus merge trick to preserve the history of the sections on lexical conventions, attributes, flags, options and tables (the ones that were rewritten the less). BTW, I definitely recommend this strategy: review first, rewrite history after (because doing the history preserving tricks early, while it is not clear whether everything is going to stay where you put it, is just a waste of time).

So this is now ready to merge. @cpitclaudel: Since Jim doesn't merge himself (for lack of a GPG key), can you take care of it?

jfehrle commented:
> Since Jim doesn't merge himself (for lack of a GPG key)

I'll try to fix that after I get the next version of Windows, due in mid-May, that uses a virtual machine to run Linux instead of emulation.  Easier to debug, then, I think.

Zimmi48 commented:
@cpitclaudel does not seem to be available. Anyone with push-access could merge this PR now? I'd like to avoid it staying open longer than it needs to, given that this is a PR moving a lot of code and thus with which more conflicts could arise that would be harder to fix than the usual documentation PR.

cpitclaudel commented:
> @cpitclaudel does not seem to be available. 

Woops. I am, but I still get drowned in messages. Will merge now.
